File: src/2025-02-24-deepclaude-rust/java/exercises/practice/state-of-tic-tac-toe/src/main/java/StateOfTicTacToe.java
class StateOfTicTacToe {
    enum GameState {
        ONGOING,
        DRAW,
        WIN
    }
    public GameState determineState(String[] board) {
        int xCount = 0, oCount = 0;
        for (String row : board) {
            for (char c : row.toCharArray()) {
                if (c == 'X') xCount++;
                else if (c == 'O') oCount++;
            }
        }

        if (xCount != oCount && xCount != oCount + 1) {
            throw new IllegalArgumentException("Invalid turn order");
        }

        boolean xWins = isWin(board, 'X');
        boolean oWins = isWin(board, 'O');

        if (xWins && oWins) {
            throw new IllegalArgumentException("Multiple winners");
        }

        if (xWins) {
            if (xCount != oCount + 1) {
                throw new IllegalArgumentException("Invalid X win");
            }
            return GameState.WIN;
        } else if (oWins) {
            if (oCount != xCount) {
                throw new IllegalArgumentException("Invalid O win");
            }
            return GameState.WIN;
        }

        if (xCount + oCount == 9) {
            return GameState.DRAW;
        }

        return GameState.ONGOING;
    }

    private boolean isWin(String[] board, char player) {
        // Check rows
        for (int i = 0; i < 3; i++) {
            if (board[i].charAt(0) == player && 
                board[i].charAt(1) == player && 
                board[i].charAt(2) == player) {
                return true;
            }
        }

        // Check columns
        for (int i = 0; i < 3; i++) {
            if (board[0].charAt(i) == player && 
                board[1].charAt(i) == player && 
                board[2].charAt(i) == player) {
                return true;
            }
        }

        // Check diagonals
        if (board[0].charAt(0) == player && 
            board[1].charAt(1) == player && 
            board[2].charAt(2) == player) {
            return true;
        }

        if (board[0].charAt(2) == player && 
            board[1].charAt(1) == player && 
            board[2].charAt(0) == player) {
            return true;
        }

        return false;
    }
}

File: src/2025-02-24-deepclaude-rust/javascript/exercises/practice/state-of-tic-tac-toe/state-of-tic-tac-toe.js
//
// This is only a SKELETON file for the 'State of Tic Tac Toe' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const gamestate = (board) => {
  // Filter out separator lines and validate row count
  const gameRows = board.filter(row => !row.includes('___'));
  if (gameRows.length !== 3) {
    throw new Error('Invalid board');
  }

  // Parse board into grid
  const grid = gameRows.map(row => {
    // Trim entire row to handle leading/trailing pipes
    const cells = row.trim().split('|').map(cell => cell.trim()).filter(cell => cell !== '');
    if (cells.length !== 3) {
      throw new Error('Invalid board');
    }
    return cells;
  });

  // Validate grid dimensions (redundant check but safe)
  if (grid.some(row => row.length !== 3)) {
    throw new Error('Invalid board');
  }

  // Count X's and O's
  let xCount = 0;
  let oCount = 0;
  for (const row of grid) {
    for (const cell of row) {
      if (cell === 'X') xCount++;
      else if (cell === 'O') oCount++;
      else if (cell !== '') {
        throw new Error('Invalid cell');
      }
    }
  }

  // Validate turn order with specific error messages
  if (xCount < oCount) {
    throw new Error('Wrong turn order: O started');
  }
  if (xCount > oCount + 1) {
    throw new Error('Wrong turn order: X went twice');
  }

  // Check for wins
  const hasWin = (player) => {
    // Check rows
    for (let i = 0; i < 3; i++) {
      if (grid[i][0] === player && grid[i][1] === player && grid[i][2] === player) {
        return true;
      }
    }

    // Check columns
    for (let i = 0; i < 3; i++) {
      if (grid[0][i] === player && grid[1][i] === player && grid[2][i] === player) {
        return true;
      }
    }

    // Check diagonals
    if (grid[0][0] === player && grid[1][1] === player && grid[2][2] === player) {
      return true;
    }
    if (grid[0][2] === player && grid[1][1] === player && grid[2][0] === player) {
      return true;
    }

    return false;
  };

  const xWin = hasWin('X');
  const oWin = hasWin('O');

  // Check for multiple winners or continued play after win
  if (xWin && oWin) {
    throw new Error('Impossible board: game should have ended after the game was won');
  }

  // Validate win conditions
  if (xWin && xCount !== oCount + 1) {
    throw new Error('Impossible board: game should have ended after the game was won');
  }
  if (oWin && oCount !== xCount) {
    throw new Error('Impossible board: game should have ended after the game was won');
  }

  // Determine game state
  if (xWin || oWin) {
    return 'win';
  } else if (xCount + oCount === 9) {
    return 'draw';
  } else {
    return 'ongoing';
  }
};
